pghoard/rohmu/object_storage/azure.py

```diff
@@ -104,47 +116,52 @@ def iter_key(self, key, *, with_metadata=True, deep=False, include_key=False):
 
     def _iter_key(self, *, path, with_metadata, deep):
         include = "metadata" if with_metadata else None
-        kwargs = {}
+        container_client = self.conn.get_container_client(self.container_name)
+        name_starts_with = None
+        delimiter = ""
         if path:
             # If you give Azure an empty path, it gives you an authentication error
-            kwargs["prefix"] = path
+            name_starts_with = path
         if not deep:
-            kwargs["delimiter"] = "/"
-        items = self.conn.list_blobs(self.container_name, include=include, **kwargs)
+            delimiter = "/"
+        items = container_client.walk_blobs(include=include, name_starts_with=name_starts_with, delimiter=delimiter)
         for item in items:
             if isinstance(item, BlobPrefix):
                 yield IterKeyItem(type=KEY_TYPE_PREFIX, value=self.format_key_from_backend(item.name).rstrip("/"))
             else:
-                if with_metadata:
+                if with_metadata and item.metadata:
                     # Azure Storage cannot handle '-' so we turn them into underscores and back again
                     metadata = {k.replace("_", "-"): v for k, v in item.metadata.items()}
                 else:
-                    metadata = None
+                    metadata = {}
```

Comment:
Is there a reason to change this, doesn't seem to be related to the SDK switch?

Reply:
There was a reason, but a bad one that made tests pass for the wrong reason. Current implementation should be correct